File: slides/02-optimization-problems/rsrc/abs.py
# -*- coding: utf-8 -*-
"""
Created on Tue Nov  2 09:23:26 2021

@author: Katharina Rath
"""

# subgradient
import numpy as np
import matplotlib.pyplot as plt
def f(x):
    return np.abs(x)
def df(x, c):
    return c*x

xmin = -1.5
xmax = 1.5
xv = np.linspace(xmin, xmax, 1000)
xgrad = np.linspace(-1, 1, 100)
cv = np.linspace(-0.9 , 0.9, 10)
xp = 0.5
yp = 0.5

labels = [r'$g_1$', r'$g_2$', r'$g_3$', r'$g_4$', r'$g_5$', r'$g_6$', r'$g_7$', r'$g_8$', r'$g_9$', r'$g_{10}$']
fig = plt.figure()
# ax = fig.add_subplot(211)
plt.plot(xv,f(xv), linewidth = 2)
plt.annotate(r'$f(x) = |x|$', xy=(-1.2, 1.2), color = 'black', fontsize = 15)
plt.xlabel('x')

# ax = fig.add_subplot(212)
cidx = 0
for c in cv:
    plt.plot(xgrad, df(xgrad, c), '--')
    plt.annotate(labels[cidx], xy=(1.05, -c), color = 'black', fontsize = 10)
    cidx = cidx + 1
# plt.xlabel('x')
# plt.ylabel(r'$\partial f(x)$')
plt.tight_layout()
plt.savefig('../figure_man/abs.png')

File: slides/02-optimization-problems/rsrc/abs.R
# ------------------------------------------------------------------------------
# optimization problems

# FIG: plot |x| and subgradients
# ------------------------------------------------------------------------------

library(ggplot2)

# DATA -------------------------------------------------------------------------

# f(x) = |x|
f <- function(x) abs(x)

# subgradient function
df <- function(x, c) c * x

xv <- seq(-1.5, 1.5, length.out = 1000)
xgrad <- seq(-1, 1, length.out = 100)
cv <- seq(-0.9, 0.9, length.out = 10)

df_f <- data.frame(x = xv, y = f(xv))

labels <- paste0("g[", 1:10, "]")
df_grad <- data.frame()
for (i in seq_along(cv)) {
  df_grad <- rbind(df_grad, data.frame(x = xgrad, y = df(xgrad, cv[i]), c = as.factor(i), label = labels[i], grad_val = -cv[i]))
}

# PLOT -------------------------------------------------------------------------

plot <- ggplot() +
  # Plot f(x) = |x|
  geom_line(data = df_f, aes(x = x, y = y), color = "black", linewidth = 1.5) +
  annotate("text", x = -1.2, y = 1, label = "f(x) == abs(x)", parse = TRUE, size = 8, color = "black") +
  
  # Plot subgradients with unique colors
  geom_line(data = df_grad, aes(x = x, y = y, color = c), linetype = "dashed", linewidth = 1) +
  
  # Annotate subgradients at (1.05, -c)
  geom_text(data = df_grad[df_grad$x == max(xgrad), ], 
            aes(x = 1.1, y = grad_val, label = label), 
            color = "black", size = 5, parse = TRUE) +
  
  # Labels and themes
  labs(x = "x", y = "", color = "g_i") +
  scale_color_manual(values = rainbow(10)) +  # Ensure unique colors
  theme_minimal() +
  theme(
    text = element_text(size = 16),
    legend.position = "none"  # Hide legend to match Python
  )

plot
ggsave("../figure_man/abs.png", plot, width = 8, height = 6)
